### PR TITLE
RTECO-1007 - Collect build-info on jf skills publish

### PR DESCRIPTION
**Pull request by udaykb2**
### Collect build-info on jf skills publish

- Support --build-name and --build-number (--module optional) on skills publish; honor JFROG_CLI_BUILD_* when flags are unset.
- Record the published skill zip as build-info partials after upload; publish with jf rt build-publish as usual.
- Default module id to the skill name from SKILL.md when --module is omitted.
- Avoided writing tests for BuildConfiguration of --build-name & --build-number, because BuildConfiguration tests for build-info are already covered in jfrog-cli-core

- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [ ] Appropriate label is added to auto generate release notes.
- [ ] I used gofmt for formatting the code before submitting the pull request.
- [ ] PR description is clear and concise, and it includes the proposed solution/fix.
-----

### CODE REVIEW

File: skills/commands/publish/publish.go [hunk 1]
```diff
@@ -478,19 +503,60 @@ func computeSHA256(path string) (string, error) {
...
+	if err != nil {
+		return nil, err
+	}
 
 	uploadParams := services.NewUploadParams()
 	uploadParams.Pattern = zipPath
 	uploadParams.Target = target
 	uploadParams.Flat = true
 
+	toCollect, err := pc.buildConfiguration.IsCollectBuildInfo()
```
Comment (itsmeleela) [inline skills/commands/publish/publish.go]: pc.buildConfiguration can never be nil here?  

I see applyDefaultBuildInfoModule function already checks this,  upload() needs the same nil check before calling IsCollectBuildInfo() right?

Comment (udaykb2) [reply skills/commands/publish/publish.go]: This nil check is not needed as IsCollectBuildInfo() is handling it.

Reason:
In Go we can call a method on a nil pointer receiver and IsCollectBuildInfo() is explicitly checks for a nil receiver up front

This is how we have used IsCollectBuildInfo() in other places in our code:
reference:
1. jfrog-cli-artifactory/artifactory/commands/yarn/yarn.go
2.jfrog-cli-artifactory/artifactory/commands/python/twine.go

But for extra layer of safety. I will add nil check from our calling side as well.

Comment (udaykb2) [reply skills/commands/publish/publish.go]: Updated and added nil check for extra safety from the calling side as well

File: skills/commands/publish/publish.go [hunk 2]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
 	return hex.EncodeToString(h.Sum(nil)), nil
 }
 
-func (pc *PublishCommand) upload(zipPath, target string) error {
-	serviceManager, err := utils.CreateUploadServiceManager(pc.serverDetails, 1, 3, 0, false, nil)
+// applyDefaultBuildInfoModule sets the build-info module id from the skill slug when
+// --module is omitted. Otherwise an empty module id is merged using the build name.
+func applyDefaultBuildInfoModule(buildCfg *build.BuildConfiguration, slug string) error {
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]: lets add a test case for buildConfiguration is nil.

File: skills/commands/publish/publish.go [hunk 3]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
 	return hex.EncodeToString(h.Sum(nil)), nil
 }
 
-func (pc *PublishCommand) upload(zipPath, target string) error {
-	serviceManager, err := utils.CreateUploadServiceManager(pc.serverDetails, 1, 3, 0, false, nil)
+// applyDefaultBuildInfoModule sets the build-info module id from the skill slug when
+// --module is omitted. Otherwise an empty module id is merged using the build name.
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]:
```suggestion
// applyDefaultBuildInfoModule sets the build-info module id from the skill slug when
// --module is omitted.
```

File: skills/commands/publish/publish.go [hunk 4]
```diff
@@ -159,10 +169,25 @@ func (pc *PublishCommand) Run() error {
 		return fmt.Errorf("failed to compute SHA256: %w", err)
 	}
 
+	if err := applyDefaultBuildInfoModule(pc.buildConfiguration, slug); err != nil {
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]: upload also can add build info. Even without in specific implementation of build info collection in below API it can auto collect, I have added 2 other suggestions look into it, depends on @shuvadipc decision of how to treat the auto deletion of skills, if build info should be captured even though there is auto deletion then reuse existing build info collection in upload functionality if not then keep custom build info collection logic.

Reference call tree where build info collection happens in upload
pc.upload(zipPath, target)
                         │
                         ├─► line 523: utils.CreateUploadServiceManager(...)
                         ├─► line 528: services.NewUploadParams()
                         ├─► line 536: pc.buildConfiguration.IsCollectBuildInfo() [build info call]
                         ├─► line 543: build.CreateBuildPropsFromConfiguration(...)
                         └─► line 549: serviceManager.UploadFilesWithSummary(...)

File: skills/commands/publish/publish.go [hunk 5]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
...
 
-func (pc *PublishCommand) upload(zipPath, target string) error {
-	serviceManager, err := utils.CreateUploadServiceManager(pc.serverDetails, 1, 3, 0, false, nil)
+// applyDefaultBuildInfoModule sets the build-info module id from the skill slug when
+// --module is omitted. Otherwise an empty module id is merged using the build name.
+func applyDefaultBuildInfoModule(buildCfg *build.BuildConfiguration, slug string) error {
+	if buildCfg == nil || buildCfg.GetModule() != "" {
+		return nil
+	}
+	toCollect, err := buildCfg.IsCollectBuildInfo()
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]: nit
```suggestion
	collectBuildInfo, err := buildCfg.IsCollectBuildInfo()
```

File: skills/commands/publish/publish.go (hunk 2 above)
Comment (bhanurp) [inline skills/commands/publish/publish.go]: nit
```suggestion
func applyDefaultBuildInfoModule(buildConfig *build.BuildConfiguration, slug string) error {
```

File: skills/commands/publish/publish.go [hunk 6]
```diff
@@ -159,10 +169,25 @@ func (pc *PublishCommand) Run() error {
 		return fmt.Errorf("failed to compute SHA256: %w", err)
 	}
 
+	if err := applyDefaultBuildInfoModule(pc.buildConfiguration, slug); err != nil {
+		return err
+	}
+
 	target := fmt.Sprintf("%s/%s/%s/", pc.repoKey, slug, version)
-	if err := pc.upload(zipPath, target); err != nil {
+	artifactsDetailsReader, err := pc.upload(zipPath, target)
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]:
```suggestion
collectBuildInfo := false
  if pc.buildConfiguration != nil {
      collectBuildInfo, err = pc.buildConfiguration.IsCollectBuildInfo()
      if err != nil {
          return err
      }
      if collectBuildInfo && pc.buildConfiguration.GetModule() == "" {
          pc.buildConfiguration.SetModule(slug)
      }
  }
	artifactsDetailsReader, err := pc.upload(zipPath, target, collectBuildInfo)
```

File: skills/commands/publish/publish.go [hunk 7]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
...
 	if err != nil {
 		return err
 	}
+	if toCollect {
+		buildCfg.SetModule(slug)
+	}
+	return nil
+}
+
+func (pc *PublishCommand) upload(zipPath, target string) (*content.ContentReader, error) {
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]:
```suggestion
func (pc *PublishCommand) upload(zipPath, target string, collectBuildInfo bool) (*content.ContentReader, error) {
```

File: skills/commands/publish/publish.go [hunk 8]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
...
 	uploadParams.Flat = true
 
+	toCollect := false
+	if pc.buildConfiguration != nil {
+		var err error
+		toCollect, err = pc.buildConfiguration.IsCollectBuildInfo()
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]:
```suggestion
```

File: skills/commands/publish/publish.go [hunk 9]
```diff
@@ -478,19 +503,64 @@ func computeSHA256(path string) (string, error) {
...
+
+	if toCollect {
```
Comment (bhanurp) [inline skills/commands/publish/publish.go]:
```suggestion
	if collectBuilfInfo {
```

**Review by bhanurp — CHANGES_REQUESTED** [review]

File: skills/commands/publish/publish.go (hunk 5 above)
Comment (udaykb2) [reply skills/commands/publish/publish.go]: done

File: skills/commands/publish/publish.go (hunk 4 above)
Comment (udaykb2) [reply skills/commands/publish/publish.go]: as discussed, this change is not required

**Review by bhanurp — APPROVED** [review]

**Review by naveenku-jfrog — APPROVED** [review]
